Cdn on edge tutorial (#52)
wasmerio/docs.wasmer.io: pages/edge/tutorials/cdn-on-wasmer-edge.mdx

import { Steps, Callout, FileTree } from "nextra-theme-docs";
import StaticSite from "@components/deploy/quickstart/StaticSite";
import edgeReactImg from "@assets/deploy/tutorials/edge-react.png";
import Install from "@components/install.mdx";
import Login from "@components/login.mdx";
import CliVersionCallout from "@components/deploy/CliVersionCallout.mdx";

> # Creating a CDN on Wasmer Edge

Reviewer: ```suggestion
# Using Wasmer Edge as a CDN
```

Author: changed that

> In this tutorial, we will create a CDN on Wasmer Edge. We'll be using the [`static-web-server`]()

Reviewer: ```suggestion
In this tutorial, we be demonstrate the CDN capabilities Wasmer Edge. For that, we'll be using the [`static-web-server`](WHERE THIS LINKS??)
```

Author: Updated that

Writing app config to '/Volumes/Work/Projects/wasmer-edge/wasmer-edge-cdn/app.yaml'
To (re)deploy your app, run 'wasmer deploy'
```

Select the above configuration mentioned in the comments and press enter.

The above command will initialize a blank project with the following structure.

<FileTree>
  <FileTree.Folder name="your-top-level-dir" defaultOpen>
    <FileTree.Folder name="public" defaultOpen>
      <FileTree.File name="index.html" />
    </FileTree.Folder>
    <FileTree.File name="app.yaml" />
    <FileTree.File name="wasmer.toml" />
  </FileTree.Folder>
</FileTree>

### Changing configuration for CDN deployment

Let's first delete the `index.html` file from the `public` directory.

```bash
$ rm public/index.html
```

Now as you might notice, this project uses the `static-web-server` package which is a static web server written in Rust.

Most notably, it supports changing the compression level, cache control, cors, and more.
All these configurations can be super useful when deploying a CDN.

So, let's add a `config.toml` file in the `config` directory.

```bash
$ mkdir config
$ touch config/config.toml
```

You can copy the following `config.toml`.

```toml
[general]

> 

Reviewer: We don't need to showcase every detail here, nor that big of a config. Please summarize it to the essential things only

# kind = 301

# [[advanced.redirects]]
# source = "/index.html"
# destination = "https://static-web-server.net"
# kind = 302

### URL Rewrites (examples only)

# [[advanced.rewrites]]
# source = "**/*.{png,ico,gif}"
# destination = "/assets/favicon.ico"
## Optional redirection
# redirect = 301

# [[advanced.rewrites]]
# source = "**/*.{jpg,jpeg}"
# destination = "/images/sws.png"

### Virtual Hosting

# [[advanced.virtual-hosts]]
## But if the "Host" header matches this...
# host = "sales.example.com"
## ...then files will be served from here instead
# root = "/var/sales/html"

# [[advanced.virtual-hosts]]
# host = "blog.example.com"
# root = "/var/blog/html"
```

<Callout type="info" emoji="ℹ️">
  The above configuration is the configuration of the `static-web-server`
  package. More info about it can be found
  [here](https://static-web-server.net/configuration/config-file/).
</Callout>

Now, let's add the following configuration to the `wasmer.toml` file.

```yaml
[package]
name = "<your-namespace>/<your-cdn-name>"
version = "0.1.0"
description = "<your-namespace>/<your-cdn-name> CDN on Wasmer Edge"

[dependencies]
"wasmer/static-web-server" = "^1"

[fs]
public = "public"

> config = "config" # 👈🏼 Add the config directory

Reviewer: We are using non absolute paths here. We should be using absolute paths

Author: I changed the paths to:
```toml
public = "/public"
config = "/config"
```

I think public is a good name for static assets ? As there's no application per-se but rather just hosting.

Reviewer: Nope, Is better to do `"/app" = .`. That way you have both `/app/public` and `/app/config.toml` automatically on it

[[command]]
name = "script"
module = "wasmer/static-web-server-async:webserver"
runner = "https://webc.org/runner/wasi"

[command.annotations.wasi] # 👈🏼 Add config.toml as an argument for the webserver

> main-args = ["-w", "./config/config.toml"]

Reviewer: Absolute paths

Author: changed that to use the path `/config/config.toml`

Reviewer: Nope, it should be `/app/config.toml`

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
  </head>
  <body>
    <img src="./image.jpg" width="600" />
  </body>
</html>
```

You can also add other files like css, js, videos, pdfs, etc.

You can view my app [here](https://dynamite-bud-sws-cdn.wasmer.app).

### Deploying the CDN

Now, let's deploy the CDN.

```bash
$ wasmer deploy
```

The above command will prompt you with the following:

```bash
Loaded app from: /path/to/your/dir/app.yaml

Publish new version of package '<your-namespace>/<your-cdn-name>'? yes
Publishing package...
[1/2] ⬆️   Uploading...
[2/2] 📦  Publishing...
Successfully published package `<your-namespace>/<your-cdn-name>@0.1.0`
Waiting for package to become available.....
Package '<your-namespace>/<your-cdn-name>@0.1.0' published successfully!

Deploying app <your-namespace>-<your-cdn-name>...

 ✅ App <your-namespace>/<your-namespace>-<your-cdn-name> was successfully deployed!

> App URL: https://dynamite-bud-sws-cdn.wasmer.app
> Versioned URL: https://2qjh9iec55z4.id.wasmer.app

> > Admin dashboard: https://wasmer.io/apps/dynamite-bud-sws-cdn

Reviewer: This url should not have anything about your user here. It should also be visitable by people

Waiting for new deployment to become available...
(You can safely stop waiting now with CTRL-C)
.
New version is now reachable at https://dynamite-bud-sws-cdn.wasmer.app
Deployment complete
```

<Callout type="info" emoji="ℹ️">
  And now you can view your CDN at the URL mentioned above. You can also view
  the admin dashboard of your CDN by visiting the URL mentioned above.
</Callout>

<Callout type="info" emoji="🌐">
  You'll need to do `wasmer deploy` again if you add any new files to the
  `public` directory.
</Callout>

<Callout type="warning" emoji="🚧">
  The url of your CDN will be different from the one mentioned above.

> </Callout>

Reviewer: There's not a "Your CDN". You don't build a CDN, you publish INTO a CDN. This needs rephrasing

Author: Rephrased to: `The url of your app will be different from the one mentioned above.`

</Steps>

## Conclusion

> In this tutorial, we learned how to create a CDN on Wasmer Edge using the `static-web-server` package.

Reviewer: No, you can't create a CDN, you use Wasmer Edge as a CDN

Author: Changed that
